src/tss2-tcti/tcti-tbs.c

```diff
@@ -0,0 +1,291 @@
+
+    tbs_rc = Tbsip_Context_Close (tcti_tbs->hContext);
+    if (tbs_rc != TBS_SUCCESS) {
+        LOG_WARNING ("Failed to close context with TBS error: 0x%x", tbs_rc);
+    }
+
+    tcti_common->state = TCTI_STATE_FINAL;
+}
+
+TSS2_RC
+tcti_tbs_cancel (
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    TBS_RESULT tbs_rc;
+    TSS2_RC rc = TSS2_RC_SUCCESS;
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs = tcti_tbs_context_cast (tctiContext);
+
+    tbs_rc = Tbsip_Cancel_Commands (tcti_tbs->hContext);
+    if (tbs_rc != TBS_SUCCESS) {
+        LOG_WARNING ("Failed to cancel commands with TBS error: 0x%x", tbs_rc);
+        rc = TSS2_TCTI_RC_GENERAL_FAILURE;
+    }
+
+    return rc;
+}
+
+TSS2_RC
+tcti_tbs_get_poll_handles (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    TSS2_TCTI_POLL_HANDLE *handles,
+    size_t *num_handles)
+{
+    /* TBS doesn't support polling. */
+    (void)(tctiContext);
+    (void)(handles);
+    (void)(num_handles);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+TSS2_RC
+tcti_tbs_set_locality (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    uint8_t locality)
+{
+    /*
+     * TBS currently only supports locality 0
+     */
+    (void)(tctiContext);
+    (void)(locality);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+TSS2_RC
+Tss2_Tcti_Tbs_Init (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t *size,
+    const char *conf)
+{
+    TSS2_TCTI_TBS_CONTEXT *tcti_tbs;
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common;
+    TBS_RESULT tbs_rc;
+    TBS_CONTEXT_PARAMS2 params;
+    TPM_DEVICE_INFO info;
+
+    if (tctiContext == NULL) {
+        if (size == NULL) {
+            return TSS2_TCTI_RC_BAD_VALUE;
+        }
+        *size = sizeof (TSS2_TCTI_TBS_CONTEXT);
+        return TSS2_RC_SUCCESS;
+    }
+
+    /* Init TCTI context */
+    TSS2_TCTI_MAGIC (tctiContext) = TCTI_TBS_MAGIC;
+    TSS2_TCTI_VERSION (tctiContext) = TCTI_VERSION;
+    TSS2_TCTI_TRANSMIT (tctiContext) = tcti_tbs_transmit;
+    TSS2_TCTI_RECEIVE (tctiContext) = tcti_tbs_receive;
+    TSS2_TCTI_FINALIZE (tctiContext) = tcti_tbs_finalize;
+    TSS2_TCTI_CANCEL (tctiContext) = tcti_tbs_cancel;
+    TSS2_TCTI_GET_POLL_HANDLES (tctiContext) = tcti_tbs_get_poll_handles;
+    TSS2_TCTI_SET_LOCALITY (tctiContext) = tcti_tbs_set_locality;
+    TSS2_TCTI_MAKE_STICKY (tctiContext) = tcti_make_sticky_not_implemented;
+    tcti_tbs = tcti_tbs_context_cast (tctiContext);
+    tcti_common = tcti_tbs_down_cast (tcti_tbs);
+    tcti_common->state = TCTI_STATE_TRANSMIT;
+
+    memset (&tcti_common->header, 0, sizeof (tcti_common->header));
+    tcti_common->locality = 0;
+
+    params.includeTpm20 = 1;
+    params.version = TBS_CONTEXT_VERSION_TWO;
+    tbs_rc = Tbsi_Context_Create ((PCTBS_CONTEXT_PARAMS)&params, &(tcti_tbs->hContext));
+    if (tbs_rc != TBS_SUCCESS) {
+        LOG_WARNING ("Failed to create context with TBS error: 0x%x", tbs_rc);
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+
+    tbs_rc = Tbsi_GetDeviceInfo (sizeof (info), &info);
+    if (tbs_rc != TBS_SUCCESS) {
```

Comment:
leaking the context created by `Tbsi_Context_Create` on the error path

Reply:
These should be fixed now, I've added calls to `Tbsip_Context_Close` on the error paths. Currently there is no check for the success of `Tbsip_Context_Close`, since `TSS2_TCTI_RC_IO_ERROR` will be returned either way. If you would like to check for the success of `Tbsip_Context_Close` for logging purposes I'll add that as well.

Reply:
I'm fine ignoring the response from the `close` function. By this point we've already logged one `WARNING` that will clearly identify the code where the failure occurred.